_posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd

```diff
@@ -0,0 +1,492 @@
+
+```{r}
+feat_type_in <- nn$FieldType(r2_act, list(r2_act$trivial_repr))
+feat_type_out <- nn$FieldType(
+  r2_act,
+  list(r2_act$regular_repr, r2_act$regular_repr, r2_act$regular_repr)
+)
+
+conv <- nn$R2Conv(feat_type_in, feat_type_out, kernel_size = 3L)
+```
+
+We then perform convolution on a batch of images, made aware of their "data type" by wrapping them in `feat_type_in`:
+
+```{r}
+x <- torch$rand(2L, 1L, 32L, 32L)
+x <- feat_type_in(x)
+y <- conv(x)
+y$shape |> unlist()
+```
+
+```         
+[1]  2  12 30 30
+```
+
+The output has twelve "channels", this being the product of group cardinality -- four distinguished positions -- and number of feature vector fields (three).
+
+If we choose the simplest possible, approximately, test case, we can verify that such a convolution is equivariant by direct inspection. Here's my setup:
+
+```{r}
+feat_type_in <- nn$FieldType(r2_act, list(r2_act$trivial_repr))
+feat_type_out <- nn$FieldType(r2_act, list(r2_act$regular_repr))
+conv <- nn$R2Conv(feat_type_in, feat_type_out, kernel_size = 3L)
+
+torch$nn$init$constant_(conv$weights, 1.)
+x <- torch$vander(torch$arange(0,4))$view(tuple(1L, 1L, 4L, 4L)) |> feat_type_in()
+x
+```
+
+```         
+g_tensor([[[[ 0.,  0.,  0.,  1.],
+            [ 1.,  1.,  1.,  1.],
+            [ 8.,  4.,  2.,  1.],
+            [27.,  9.,  3.,  1.]]]], [C4_on_R2[(None, 4)]: {irrep_0 (x1)}(1)])
+```
+
+Inspection could be performed using any group element. I'll pick rotation by $\pi/2$:
+
+```{r}
+all <- iterate(r2_act$testing_elements)
+g1 <- all[[2]]
+g1
+```
+
+Just for fun, let's see how we can -- literally -- come whole circle by letting this element act on the input tensor four times:
+
+```{r}
+all <- iterate(r2_act$testing_elements)
+g1 <- all[[2]]
+
+x1 <- x$transform(g1)
+x1$tensor
+x2 <- x1$transform(g1)
+x2$tensor
+x3 <- x2$transform(g1)
+x3$tensor
+x4 <- x3$transform(g1)
+x4$tensor
+```
+
+```         
+tensor([[[[ 1.,  1.,  1.,  1.],
+          [ 0.,  1.,  2.,  3.],
+          [ 0.,  1.,  4.,  9.],
+          [ 0.,  1.,  8., 27.]]]])
+          
+tensor([[[[ 1.,  3.,  9., 27.],
+          [ 1.,  2.,  4.,  8.],
+          [ 1.,  1.,  1.,  1.],
+          [ 1.,  0.,  0.,  0.]]]])
+          
+tensor([[[[27.,  8.,  1.,  0.],
+          [ 9.,  4.,  1.,  0.],
+          [ 3.,  2.,  1.,  0.],
+          [ 1.,  1.,  1.,  1.]]]])
+          
+tensor([[[[ 0.,  0.,  0.,  1.],
+          [ 1.,  1.,  1.,  1.],
+          [ 8.,  4.,  2.,  1.],
+          [27.,  9.,  3.,  1.]]]])
+```
+
+You see that at the end, we are back at the original "image".
+
+Now, for equivariance. We could first apply a rotation, then convolve. Rotate:
+
+```{r}
+x_rot <- x$transform(g1)
+x_rot$tensor
+```
+
+This is the first in the above list of four tensors. -- Convolve:
```

Comment:
I'm not sure what the relationship between the words/sentences is meant to be when you have a period followed by an em dash and a word with a capital letter, e.g.,
> . -- Convolve
I'm not suggesting anything else, because I don't know what you're going for.